rust/otap-dataflow/crates/otap/src/experimental/condense_attributes_processor/mod.rs

```diff
@@ -0,0 +1,1112 @@
+            .as_any()
+            .downcast_ref::<DictionaryArray<UInt16Type>>(
+        ) {
+            let vals = dict
+                .values()
+                .as_any()
+                .downcast_ref::<StringArray>()
+                .ok_or_else(|| {
+                    engine_err("dictionary values are not StringArray for UInt16 key type")
+                })?;
+            (Some(dict.keys()), Some(vals))
+        } else {
+            (None, None)
+        };
+
+        // Helper function to get key at index i
+        let get_key = |i: usize| -> Option<&str> {
+            if let Some(arr) = key_str_arr {
+                return if arr.is_null(i) {
+                    None
+                } else {
+                    Some(arr.value(i))
+                };
+            }
+            if let (Some(keys), Some(vals)) = (key_dict_u8_keys, key_dict_u8_vals) {
+                return if keys.is_null(i) {
+                    None
+                } else {
+                    Some(vals.value(keys.value(i) as usize))
+                };
+            }
+            if let (Some(keys), Some(vals)) = (key_dict_u16_keys, key_dict_u16_vals) {
+                return if keys.is_null(i) {
+                    None
+                } else {
+                    Some(vals.value(keys.value(i) as usize))
+                };
+            }
+            None
+        };
+
+        // Helper function to extract value as string based on type
+        let get_value_str = |value_type: AttributeValueType, i: usize| -> Option<String> {
+            match value_type {
+                AttributeValueType::Str => str_col.and_then(|col| {
+                    Self::extract_value_from_column(col, i, |arr: &StringArray, index| {
+                        arr.value(index).to_string()
+                    })
+                }),
+                AttributeValueType::Int => int_col.and_then(|col| {
+                    Self::extract_value_from_column(col, i, |arr: &Int64Array, index| {
+                        arr.value(index)
+                    })
+                    .map(|v| v.to_string())
+                }),
+                AttributeValueType::Double => double_col.and_then(|col| {
+                    Self::extract_value_from_column(col, i, |arr: &Float64Array, index| {
+                        arr.value(index)
+                    })
+                    .map(|v| v.to_string())
+                }),
+                AttributeValueType::Bool => bool_col.and_then(|col| {
+                    Self::extract_value_from_column(col, i, |arr: &BooleanArray, index| {
+                        arr.value(index)
+                    })
+                    .map(|v| v.to_string())
+                }),
+                // If needed, add handling for Map, Slice, and Bytes?
+                _ => None,
+            }
+        };
+
+        let mut parent_to_attrs: HashMap<u16, Vec<(&str, String)>> = HashMap::new();
+        let mut rows_to_preserve: Vec<usize> = Vec::new();
```

Comment:
More of the optimization suggestion (not blocker) - `parent_to_attrs` and `rows_to_preserve` are allocated on every `condense()` call. We can consider moving them to the processor struct and using `.clear()` between batches instead of reallocating

Reply:
Yes i thought about this too, probably 'String' would still have better performance. But right to leave it for now as this would need comparing perf results before deciding.